test/integration/controllermanager/cloudprofile/cloudprofile_test.go

```diff
@@ -0,0 +1,168 @@
+				Regions: []gardencorev1beta1.Region{
+					{Name: "some-region"},
+				},
+			},
+		}
+		shoot = &gardencorev1beta1.Shoot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      resourceName,
+				Namespace: testNamespace.Name,
+			},
+			Spec: gardencorev1beta1.ShootSpec{
+				CloudProfileName:  cloudProfile.Name,
+				SecretBindingName: "my-provider-account",
+				Region:            "foo-region",
+				Provider: gardencorev1beta1.Provider{
+					Type: cloudProfile.Spec.Type,
+					Workers: []gardencorev1beta1.Worker{
+						{
+							Name:    "cpu-worker",
+							Minimum: 2,
+							Maximum: 2,
+							Machine: gardencorev1beta1.Machine{Type: "large"},
+						},
+					},
+				},
+				Kubernetes: gardencorev1beta1.Kubernetes{Version: "1.21.1"},
+				Networking: gardencorev1beta1.Networking{Type: "foo-networking"},
+			},
+		}
+	})
+
+	JustBeforeEach(func() {
+		By("Create CloudProfile")
+		Expect(testClient.Create(ctx, cloudProfile)).To(Succeed())
+		log.Info("Created CloudProfile for test", "cloudProfile", client.ObjectKeyFromObject(cloudProfile))
+
+		DeferCleanup(func() {
+			By("Delete CloudProfile")
+			Expect(client.IgnoreNotFound(testClient.Delete(ctx, cloudProfile))).To(Succeed())
+		})
+
+		if shoot != nil {
+			By("Create Shoot")
+			Expect(testClient.Create(ctx, shoot)).To(Succeed())
+			log.Info("Created shoot for test", "shoot", client.ObjectKeyFromObject(shoot))
+
+			DeferCleanup(func() {
+				By("Delete Shoot")
+				Expect(client.IgnoreNotFound(gardener.ConfirmDeletion(ctx, testClient, shoot))).To(Succeed())
+				Expect(client.IgnoreNotFound(testClient.Delete(ctx, shoot))).To(Succeed())
+			})
+		}
+	})
+
+	Context("no shoot referencing the CloudProfile", func() {
+		BeforeEach(func() {
+			shoot = nil
+		})
+
+		It("should add the finalizer and release it on deletion", func() {
+			By("Ensure finalizer got added")
+			Eventually(func(g Gomega) {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(cloudProfile), cloudProfile)).To(Succeed())
+				g.Expect(cloudProfile.Finalizers).To(ConsistOf("gardener"))
+			}).Should(Succeed())
+
+			By("Delete CloudProfile")
+			Expect(testClient.Delete(ctx, cloudProfile)).To(Succeed())
+
+			By("Ensure finalizer got removed")
+			Eventually(func() error {
+				return testClient.Get(ctx, client.ObjectKeyFromObject(cloudProfile), cloudProfile)
+			}).Should(BeNotFoundError())
+		})
+	})
+
+	Context("shoots referencing the CloudProfile", func() {
+		JustBeforeEach(func() {
+			By("Ensure finalizer got added")
+			Eventually(func(g Gomega) {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(cloudProfile), cloudProfile)).To(Succeed())
+				g.Expect(cloudProfile.Finalizers).To(ConsistOf("gardener"))
+			}).Should(Succeed())
+
+			By("Delete CloudProfile")
+			Expect(testClient.Delete(ctx, cloudProfile)).To(Succeed())
+		})
+
+		It("should add the finalizer and not release it on deletion since there still is a referencing shoot", func() {
+			By("Ensure finalizer got removed")
```

Comment:
Comment in `By` is a bit confusing, It says `Ensure finalizer got removed` but we are checking if the cloudprofile is still there or not.